File: clients/cpp/model/WeaponParams.cpp
#include "WeaponParams.hpp"

WeaponParams::WeaponParams() { }
WeaponParams::WeaponParams(int magazineSize, double fireRate, double reloadTime, double minSpread, double maxSpread, double recoil, double aimSpeed, BulletParams bullet, std::shared_ptr<ExplosionParams> explosion) : magazineSize(magazineSize), fireRate(fireRate), reloadTime(reloadTime), minSpread(minSpread), maxSpread(maxSpread), recoil(recoil), aimSpeed(aimSpeed), bullet(bullet), explosion(explosion) { }
WeaponParams WeaponParams::readFrom(InputStream& stream) {
    WeaponParams result;
    result.magazineSize = stream.readInt();
    result.fireRate = stream.readDouble();
    result.reloadTime = stream.readDouble();
    result.minSpread = stream.readDouble();
    result.maxSpread = stream.readDouble();
    result.recoil = stream.readDouble();
    result.aimSpeed = stream.readDouble();
    result.bullet = BulletParams::readFrom(stream);
    if (stream.readBool()) {
        result.explosion = std::shared_ptr<ExplosionParams>(new ExplosionParams());
        *result.explosion = ExplosionParams::readFrom(stream);
    } else {
        result.explosion = std::shared_ptr<ExplosionParams>();
    }
    return result;
}
void WeaponParams::writeTo(OutputStream& stream) const {
    stream.write(magazineSize);
    stream.write(fireRate);
    stream.write(reloadTime);
    stream.write(minSpread);
    stream.write(maxSpread);
    stream.write(recoil);
    stream.write(aimSpeed);
    bullet.writeTo(stream);
    if (explosion) {
        stream.write(false);
    } else {
        stream.write(true);
        (*explosion).writeTo(stream);
    }
}
std::string WeaponParams::toString() const {
    return std::string("WeaponParams") + "(" +
        std::to_string(magazineSize) +
        std::to_string(fireRate) +
        std::to_string(reloadTime) +
        std::to_string(minSpread) +
        std::to_string(maxSpread) +
        std::to_string(recoil) +
        std::to_string(aimSpeed) +
        bullet.toString() +
        "TODO" + 
        ")";
}

File: clients/go/model/weapon_params.go
package model

import (
	"io"

	mStream "../stream"
)

type WeaponParams struct {
	MagazineSize int32
	FireRate     float64
	ReloadTime   float64
	MinSpread    float64
	MaxSpread    float64
	Recoil       float64
	AimSpeed     float64
	Bullet       *BulletParams
	Explosion    *ExplosionParams
}

func NewWeaponParams(magazineSize int32, fireRate float64, reloadTime float64, minSpread float64,
	maxSpread float64, recoil float64, aimSpeed float64, bullet *BulletParams, explosion *ExplosionParams) *WeaponParams {
	return &WeaponParams{
		MagazineSize: magazineSize,
		FireRate:     fireRate,
		ReloadTime:   reloadTime,
		MinSpread:    minSpread,
		MaxSpread:    maxSpread,
		Recoil:       recoil,
		AimSpeed:     aimSpeed,
		Bullet:       bullet,
		Explosion:    explosion,
	}
}
func ReadWeaponParams(reader io.Reader) *WeaponParams {
	result := &WeaponParams{
		MagazineSize: mStream.ReadInt32(reader),
		FireRate:     mStream.ReadFloat64(reader),
		ReloadTime:   mStream.ReadFloat64(reader),
		MinSpread:    mStream.ReadFloat64(reader),
		MaxSpread:    mStream.ReadFloat64(reader),
		Recoil:       mStream.ReadFloat64(reader),
		AimSpeed:     mStream.ReadFloat64(reader),
		Bullet:       ReadBulletParams(reader),
	}
	if mStream.ReadBool(reader) {
		result.Explosion = ReadExplosionParams(reader)
	} else {
		result.Explosion = nil
	}
	return result
}
func (value WeaponParams) Write(writer io.Writer) {
	mStream.WriteInt32(writer, value.MagazineSize)
	mStream.WriteFloat64(writer, value.FireRate)
	mStream.WriteFloat64(writer, value.ReloadTime)
	mStream.WriteFloat64(writer, value.MinSpread)
	mStream.WriteFloat64(writer, value.MaxSpread)
	mStream.WriteFloat64(writer, value.Recoil)
	mStream.WriteFloat64(writer, value.AimSpeed)
	value.Bullet.Write(writer)
	if value.Explosion == nil {
		mStream.WriteBool(writer, false)
	} else {
		mStream.WriteBool(writer, true)
		(*value.Explosion).Write(writer)
	}
}
